### PR TITLE
KAFKA-16106: revert classic state transitions if deletion fails

### PR DESCRIPTION
**Pull request by jeffkbkim**
An expire-group-metadata operation generates tombstone records, updates the groups state and decrements group size counters, then performs a write to the log. If there is a __consumer_offsets partition reassignment, for instance, this operation fails. The groups state is reverted to an earlier snapshot but classic group size counters are not. This begins an inconsistency between the metrics and the actual groups size. This applies to all unsuccessful write operations that alter the classic group state.

However, some operations that alter the classic group state does not produce records. This means that we cannot rely on timeline data structures as we do for consumer group states.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java [hunk 1]
```diff
@@ -575,18 +581,32 @@ public CoordinatorResult<OffsetDeleteResponseData, CoordinatorRecord> deleteOffs
     public CoordinatorResult<Void, CoordinatorRecord> cleanupGroupMetadata() {
         long startMs = time.milliseconds();
         List<CoordinatorRecord> records = new ArrayList<>();
+        AtomicInteger deletedClassicGroupCount = new AtomicInteger(0);
```
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java:580]: Just to confirm, is the `AtomicInteger` used to match the following forEach loop?

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorShardTest.java [hunk 2]
```diff
@@ -993,13 +998,50 @@ public void testCleanupGroupMetadata() {
...
+            mockTime,
+            timer,
+            mock(GroupCoordinatorConfig.class),
+            mock(CoordinatorMetrics.class),
+            metricsShard
+        );
+
+        when(groupMetadataManager.groupIds()).thenReturn(mkSet("group-id", "other-group-id"));
+        when(offsetMetadataManager.cleanupExpiredOffsets(eq("group-id"), eq(new ArrayList<>()))).thenReturn(true);
+        when(groupMetadataManager.maybeDeleteGroup(eq("group-id"), eq(new ArrayList<>()))).thenReturn(true);
```
Comment (dongnuo123) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorShardTest.java:1021]: Should we add something to the record list and assert it's non null later?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java (hunk 1 above)
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java:580]: If you mean by using it to conform to the lambda expression, yes

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorShardTest.java (hunk 2 above)
Comment (jeffkbkim) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorShardTest.java:1021]: the test above, testCleanupGroupMetadata (https://github.com/apache/kafka/pull/16511/files/b73af9c786d4ad29259d0bfeb7c16db3324eff4b#diff-3a0b9cad0253e0f6d4665efd0d6f7efd5bd5dd96d3ba31005cab06fa728aad8fR990)

tests that the records we add are reflected. would this be sufficient?

**Comment by jeffkbkim** [comment]
closing and reopening with https://github.com/apache/kafka/pull/16874

**Pull request closed without merging**